docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

```diff
@@ -35,11 +38,103 @@ External patches have the following advantages:
+          description: "proxy for http calls."
+    # External patch definitions.
+    patches:
+    - name: lbImageRepository
+      external:
+          generateExtension: generate-patches.k8s-upgrade-with-runtimesdk
+          validateExtension: validate-topology.k8s-upgrade-with-runtimesdk
+          ## Explicitly call variable discovery for this patch.
+          variableDiscovery: variable-discovery.k8s-upgrade-with-runtimesdk
+status:
+    # observedGeneration is used to check that the current version of the ClusterClass is the same as that when the Status was previously written.
+    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
+    observedGeneration: xx
+    # variables contains a list of all variable definitions, both inline and external, that belong to the ClusterClass.
+    variables: 
+        - name: http-proxy
+          schema:
+            openAPIV3Schema:
+             type: string
+             # default is different so this variable does not match the inline definition.
+             default: "proxy.example3.com"
+             example: "proxy.example.com"
+             description: "proxy for http calls."
+          # namespace is added as this variable's schema isn't equivalent to the inline definition.
+          namespace: lbImageRepository
+          # A variable can have multiple origins (altenate name /sources/??)
+          # TODO: What purpose does this serve?
+         # origins: 
+         #  VariableDiscovery:
+         #  - patch: lbImageRepository # not sure if to use this or variableDiscovery << to list
+        - name: http-proxy
+          schema:
+          openAPIV3Schema:
+            type: string
+            default: "proxy.example.com"
+            example: "proxy.example.com"
+            description: "proxy for http calls."
+          namespace: inline
+```
+
+### Variable namespacing
+Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks.
+If all variables that share a name have equivalent schemas the variables are considered to be in the global namespace.
+Note: This is always the case if there is only one variable with a particular name.
+The CAPI components will consider variable definitions to be equivalent when they share a name and `openAPIV3Schema` fields
+excluding `description` and `example` are exactly equal. (TODO: Clarify this.)
+
+If two or more variables which share a name do not have equivalent schemas, the variables will be namespaced. This means they will
+appear namespaced in the ClusterClass `.status.variables` and setting values in the Cluster `.topology.spec.variables` will also require setting a namespace.
+The namespace on a variable will depend on its source (TODO: RUNTIME_EXTENSION name or patch name?)
+Note: When namespaced inline variables can be set without a namespace as they are always in the global namespace.
```

Comment:
I'm not sure if I would do this. This makes transitions from global<=>namespaced and back very magical

I think I would keep it simple and say:
* if a variable definition is namespaced variable values must be as well
* if a variable is global, namespace won't be set in the ClusterClass.status variable definitions and in the variable values in Cluster.spec.topology.variables

Reply:
+1 to Stefan's suggestion. That would make it simple. 
All inline variables implicitly becoming global is confusing especially if there are clashes.

Reply:
Small correction: 

> if a variable is global, namespace won't be set in the ClusterClass.status variable definitions and in the variable values in Cluster.spec.topology.variables

Not sure about if we should set the namespace in the ClusterClass.status.variable regardless. Probably it makes sense to always set it there to keep it simpler (and less magical)